src/ast_editor.ml

```diff
@@ -0,0 +1,585 @@
+    | Unknown
+
+  let relative_document_path ~document =
+    Workspace.asRelativePath ~pathOrUri:(`Uri (TextDocument.uri document)) ()
+
+  let project_root_path () = Workspace.rootPath ()
+
+  let get_kind ~document =
+    let relative = relative_document_path ~document in
+    match Caml.Filename.extension relative with
+    | ".ml" -> Structure
+    | ".mli" -> Signature
+    | _ -> Unknown
+
+  let get_pp_path ~(document : TextDocument.t) =
+    let relative = relative_document_path ~document in
+    match project_root_path () with
+    | None ->
+      let (_ : _ Promise.t) =
+        Window.showErrorMessage ~message:"Project root wasn't found." ()
+      in
+      None
+    | Some root ->
+      let build_root = "_build/default" in
+      let fname =
+        match get_kind ~document with
+        | Unknown ->
+          let uri = Uri.toString (TextDocument.uri document) () in
+          raise (User_error (sprintf "File %s has unknown file extension" uri))
+        | Structure -> String.chop_suffix_exn ~suffix:".ml" relative ^ ".pp.ml"
+        | Signature ->
+          String.chop_suffix_exn ~suffix:".mli" relative ^ ".pp.mli"
+      in
+      let ( / ) = Caml.Filename.concat in
+      Some (root / build_root / fname)
+end
+
+let fetch_pp_code ~document =
+  match Pp_path.get_pp_path ~document with
+  | None ->
+    show_message `Error "Unable to find preprocessed file for %s"
+      (Uri.toString (TextDocument.uri document) ());
+    Ok ""
+  | Some path -> (
+    match Ppx_tools.get_reparsed_code_from_pp_file ~path with
+    | Ok code -> Ok code
+    | Error (Io_error err_msg) -> Error err_msg
+    | Error (Read_error err_msg) ->
+      show_message `Error "%s" err_msg;
+      Ok "")
+
+let transform_to_ast instance ~(document : TextDocument.t)
+    ~(webview : WebView.t) =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let module Dumpast = Ppx_tools.Dumpast in
+  let make_value result =
+    match result with
+    | Ok ast -> ast
+    | Error error_msg -> raise (User_error error_msg)
+  in
+  if Ast_editor_state.get_original_mode ast_editor_state then
+    let origin_json =
+      let text = TextDocument.getText document () in
+      match Pp_path.get_kind ~document with
+      | Structure -> make_value (Dumpast.transform text `Impl)
+      | Signature -> make_value (Dumpast.transform text `Intf)
+      | Unknown -> raise (User_error "Unknown file extension")
+    in
+    send_msg "ast" (Jsonoo.t_to_js origin_json) ~webview
+  else
+    let pp_value =
+      match Pp_path.get_pp_path ~document with
+      | None -> raise (User_error "Project root path wasn't found")
+      | Some path -> (
+        match Ppx_tools.get_preprocessed_ast path with
+        | Error (Io_error err_msg)
+        | Error (Read_error err_msg) ->
+          raise (User_error err_msg)
+        | Ok res ->
+          let pp_json =
+            let pp_code =
+              match fetch_pp_code ~document with
+              | Ok s -> s
```

Comment:
When `s = ""` as a result of not being able to parse the file, why are we trying to parse it again?

I think `fetch_pp_code` should return a more compound type and let the callers report errors.
```ocaml
val fetch_pp_code : TextDocument.t -> (string, Ppx_tools.read_error) result
```

This will also prevent some errors being displayed both in the webview and in the popup.

Reply:
7a6c155dc2f8adae4940884cb1eb027e6a980815

Reply:
It seems like the `read_error` type isn't useful anymore. As far as I can tell, `Io_error` and `Read_error` are now handled exactly the same way. So we can drop the distinction if that's intended.